gcc/rust/typecheck/rust-hir-type-check-pattern.cc

```diff
@@ -43,32 +43,97 @@ TypeCheckPattern::Resolve (HIR::Pattern *pattern, TyTy::BaseType *parent)
 void
 TypeCheckPattern::visit (HIR::PathInExpression &pattern)
 {
-  infered = TypeCheckExpr::Resolve (&pattern);
-
-  /*
-   * We are compiling a PathInExpression, which can't be a Struct or Tuple
-   * pattern. We should check that the declaration we are referencing IS NOT a
-   * struct pattern or tuple with values.
-   */
+  // Pattern must be enum variants, sturcts, constants, or associated constansts
+  TyTy::BaseType *pattern_ty = TypeCheckExpr::Resolve (&pattern);
+
+  NodeId ref_node_id = UNKNOWN_NODEID;
+  bool maybe_item = false;
+  maybe_item
+    |= resolver->lookup_resolved_name (pattern.get_mappings ().get_nodeid (),
+				       &ref_node_id);
+  maybe_item
+    |= resolver->lookup_resolved_type (pattern.get_mappings ().get_nodeid (),
+				       &ref_node_id);
+
+  if (maybe_item)
+    {
```

Comment:
If I understand correctly, lookup_resolved_type will fail if we need to do type inferencing. Is that correct?

Reply:
AFAIK resolving (variable names and type names) is done before lowering AST into HIR.

Reply:
This section in rustc guide might be useful.
https://rustc-dev-guide.rust-lang.org/overview.html#lexing-and-parsing